urlbf4fd38f68_rduerig_StrategyBDD_tgz-pJ8-com_strategy_util_TestMainJ8/src/com/strategy/havannah/board/BoardHavannah.java

```diff
@@ -0,0 +1,336 @@
+		for (int i = 0; i < getRows(); i++) {
+			for (int k = 0; k < Math.abs(getBoardSize() - 1 - i) + 2; k++) {
+				sb.append("  ");
+			}
+			for (int j = linestart(i); j < lineend(i); j++) {
+				if (isValidField(PositionHexagon.get(i, j))) {
+					Field field = getField(i, j);
+					sb.append("|" + field.getPosition().getRow() + ":"
+							+ field.getPosition().getCol() + "|");
+				}
+			}
+			sb.append("\n");
+		}
+
+		return sb.toString();
+	}
+
+	@Override
+	public String toRowConstantString() {
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < getRows(); i++) {
+			for (int k = 0; k < Math.abs(getBoardSize() - 1 - i) + 2; k++) {
+				sb.append("  ");
+			}
+			for (int j = linestart(i); j < lineend(i); j++) {
+				if (isValidField(PositionHexagon.get(i, j))) {
+					Field field = getField(i, j);
+					RowConstant coord = RowConstant.parse(field.getIndex(),
+							getBoardSize());
+					Integer coordNumber = RowConstant.parseToCoordNumber(
+							field.getIndex(), getBoardSize());
+					sb.append(coord.name() + coordNumber + "  ");
+				}
+			}
+			sb.append("\n");
+		}
+
+		return sb.toString();
+	}
+
+	@Override
+	public String toRatingString(double[] rating, int bestIndex) {
+		EvaluationFormatter formatter = new EvaluationFormatter(bestIndex);
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < getRows(); i++) {
+			for (int k = 0; k < Math.abs(getBoardSize() - 1 - i) + 2; k++) {
+				sb.append(formatter.space());
+			}
+			for (int j = linestart(i); j < lineend(i); j++) {
+				if (isValidField(PositionHexagon.get(i, j))) {
+					Field field = getField(i, j);
+					Integer index = field.getIndex();
+					sb.append(formatter.format(index, rating[index]));
+				}
+			}
+			sb.append("\n");
+		}
+
+		return sb.toString();
+	}
+
+	// ************************************************************************
+
+	private void init(Map<Position, Integer> board) {
+		for (Entry<Position, Integer> entry : board.entrySet()) {
+			Position pos = entry.getKey();
+			int index = getIndex(pos);
+			Field field = FieldGenerator.create(entry.getValue(), pos, index);
+			fields.put(field.getPosition(), field);
+			hgfCoordinates.put(RowConstant.parse(index, boardSize),
+					RowConstant.parseToCoordNumber(index, boardSize), field);
+		}
+		System.out.println(hgfCoordinates);
+	}
+
+	private int getIndex(Position p) {
+		return getRows() * p.getRow() + p.getCol();
+	}
+
+	private void init(int[][] board, int boardSize) {
+		int limit = ((2 * boardSize) - 1);
+
+		for (int i = 0; i < limit; i++) {
+			if (i < boardSize - 1) {
+				for (int j = 0; j < ((boardSize + i) % limit); j++) {
+					Position pos = PositionHexagon.get(i, j);
+					int index = getRows() * i + j;
+					Field field = FieldGenerator
+							.create(board[i][j], pos, index);
+					fields.put(pos, field);
+					hgfCoordinates.put(RowConstant.parse(index, boardSize),
+							RowConstant.parseToCoordNumber(index, boardSize),
+							field);
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 301:310, 322:333.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.